File: macos/Sources/AeroPointAgent/Pairing/PairingService.swift
import Foundation

public struct PairingSession: Equatable, Sendable {
    public let nonce: String
    public let payload: String
}

public enum PairingError: Error, Equatable, Sendable {
    case invalidNonce
}

public protocol PairingTokenStore: AnyObject {
    func save(token: String, for clientID: String)
    func token(for clientID: String) -> String?
}

public final class InMemoryPairingTokenStore: PairingTokenStore {
    private var tokensByClientID: [String: String] = [:]

    public init() {}

    public func save(token: String, for clientID: String) {
        tokensByClientID[clientID] = token
    }

    public func token(for clientID: String) -> String? {
        tokensByClientID[clientID]
    }
}

public final class PairingService {
    private let host: String
    private let port: Int
    private let serverName: String
    private let tokenStore: PairingTokenStore
    private let nonceGenerator: () -> String
    private let tokenGenerator: () -> String
    private var activeNonce: String?

    public init(
        host: String,
        port: Int,
        serverName: String,
        tokenStore: PairingTokenStore = InMemoryPairingTokenStore(),
        nonceGenerator: @escaping () -> String = { UUID().uuidString },
        tokenGenerator: @escaping () -> String = { UUID().uuidString }
    ) {
        self.host = host
        self.port = port
        self.serverName = serverName
        self.tokenStore = tokenStore
        self.nonceGenerator = nonceGenerator
        self.tokenGenerator = tokenGenerator
    }

    public func startPairing() -> PairingSession {
        let nonce = nonceGenerator()
        activeNonce = nonce
        // Save the nonce as a valid token immediately under the pairing sentinel key.
        // The iPhone sends the nonce as its token in the hello message, so the server
        // can authenticate it before completePairing() is called.
        tokenStore.save(token: nonce, for: PairingService.pairingClientId)
        return PairingSession(
            nonce: nonce,
            payload: "aeropoint://pair?host=\(host)&port=\(port)&nonce=\(nonce)&name=\(serverName)&v=1"
        )
    }

    /// The sentinel clientId used for the initial nonce-based authentication.
    public static let pairingClientId = "__pairing__"

    public func completePairing(nonce: String, clientID: String) throws -> String {
        guard nonce == activeNonce else {
            throw PairingError.invalidNonce
        }

        let token = tokenGenerator()
        tokenStore.save(token: token, for: clientID)
        activeNonce = nil
        return token
    }
}

File: windows/Pairing/PairingService.cs
using System;
using AeroPointAgent.Storage;

namespace AeroPointAgent.Pairing
{
    public struct PairingSession
    {
        public string Nonce { get; }
        public string Payload { get; }

        public PairingSession(string nonce, string payload)
        {
            Nonce = nonce;
            Payload = payload;
        }
    }

    public sealed class PairingService
    {
        private readonly string _host;
        private readonly int _port;
        private readonly string _serverName;
        private readonly IPairingTokenStore _tokenStore;
        private readonly Func<string> _nonceGenerator;
        private readonly Func<string> _tokenGenerator;
        private string? _activeNonce;

        public const string PairingClientId = "__pairing__";

        public PairingService(
            string host,
            int port,
            string serverName,
            IPairingTokenStore tokenStore,
            Func<string>? nonceGenerator = null,
            Func<string>? tokenGenerator = null)
        {
            _host = host;
            _port = port;
            _serverName = serverName;
            _tokenStore = tokenStore;
            _nonceGenerator = nonceGenerator ?? (() => Guid.NewGuid().ToString());
            _tokenGenerator = tokenGenerator ?? (() => Guid.NewGuid().ToString());
        }

        public PairingSession StartPairing()
        {
            string nonce = _nonceGenerator();
            _activeNonce = nonce;
            // Save the nonce as a valid token immediately under the pairing sentinel key.
            // The iPhone sends the nonce as its token in the hello message, so the server
            // can authenticate it before completePairing() is called.
            _tokenStore.Save(nonce, PairingClientId);

            string payload = $"aeropoint://pair?host={_host}&port={_port}&nonce={nonce}&name={Uri.EscapeDataString(_serverName)}&v=1";
            return new PairingSession(nonce, payload);
        }

        public string CompletePairing(string nonce, string clientId)
        {
            if (nonce != _activeNonce)
            {
                throw new InvalidOperationException("Invalid pairing nonce");
            }

            string token = _tokenGenerator();
            _tokenStore.Save(token, clientId);
            _activeNonce = null;
            return token;
        }
    }
}
